### PR TITLE
#377 Codec parser fixed

### PR DESCRIPTION
**Pull request by JaylinYu**

### CODE REVIEW

**Review by alvin1221 — CHANGES_REQUESTED** [review]

**Review by JaylinYu — COMMENTED** [review]
THANKS

**Review by alvin1221 — APPROVED** [review]

File: nanomq/sub_handler.c [hunk 1]
```diff
@@ -336,15 +275,20 @@ sub_ctx_handle(nano_work *work)
 			}
 			tq = tq->next;
 		}
-		if (!topic_exist && topic_str) {
-			// printf("Protocol Version: %d\n",
-			// old_ctx->proto_ver); printf("sub ID: %d\n",
-			// work->sub_pkt->sub_id.varint);
-			int          t = work->sub_pkt->sub_id.varint;
+		if (!topic_exist) {
```
Comment (wanghaEMQ) [inline nanomq/sub_handler.c:278]: It should be ``` !topic_exist && topic_str ```

File: nanomq/sub_handler.c [hunk 2]
```diff
@@ -336,15 +275,20 @@ sub_ctx_handle(nano_work *work)
...
+				if (pdata) {
+					t = pdata->p_value.varint;
+				}
+			}
 			dbtree_ctxt *db_old_ctxt = dbtree_new_ctxt(old_ctx, t);
 			dbtree_insert_client(
 			    work->db, topic_str, db_old_ctxt, work->pid.id);
-
 			dbhash_insert_topic(work->pid.id, topic_str);
 		}
```
Comment (wanghaEMQ) [inline nanomq/sub_handler.c:294]: Here we need to reset variable ```r``` to NULL.

File: nanomq/include/pub_handler.h [hunk 3]
```diff
@@ -24,69 +24,32 @@ struct fixed_header {
...
 	struct {
-		uint16_t           packet_identifier;
+		uint16_t           packet_id;
 		struct mqtt_string topic_name;
-		struct properties  properties;
+		property           *properties;
+		uint32_t           prop_len;
 	} publish;
 
 	struct {
```
Comment (lee-emqx) [inline nanomq/include/pub_handler.h:36]: If it will increase readability declare struct out of this union ？

**Review by lee-emqx — APPROVED** [review]

**Comment by alvin1221** [comment]
Here's the log about memory leaks:
```bash
Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: tcptran_pipe_recv_cb 0x6130000202c0

Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: new 2 recevied 2 header e0 0 pos: 2 len : 0
Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: still need byte count:0 > 0

Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: pipe 0x6130000202c0 header got: e0 0 0 4 4d, 2!!

Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: remain_len 0 cparam 0x611000040000 clientid nanomq-053d63bf username (null) proto 5

Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: The type of msg is e0
Fri Mar 11 21:48:02 2022 nni_aio_finish_sync: Hunky dory(0)
Fri Mar 11 21:48:02 2022 nano_pipe_recv_cb: ######### nano_pipe_recv_cb ############
Fri Mar 11 21:48:02 2022 nano_pipe_close:  ############## nano_pipe_close ############## 
Fri Mar 11 21:48:02 2022 nni_aio_finish_error: Object closed(7)
Fri Mar 11 21:48:02 2022 nni_aio_finish_sync: Hunky dory(0)
Fri Mar 11 21:48:02 2022 server_cb: RECV  ^^^^ ctx16 ^^^^

Fri Mar 11 21:48:02 2022 decode_pub_message: cmd: 3, retain: 0, qos: 0, dup: 0, remaining length: 112
Fri Mar 11 21:48:02 2022 decode_pub_message: topic: [$SYS/brokers/disconnected], qos: 0
Fri Mar 11 21:48:02 2022 decode_pub_message: used pos: [27]
Fri Mar 11 21:48:02 2022 decode_pub_message: payload: [{"username":"(null)","ts":1052540307,"reason_code":"0","client_id":"nanomq-053d63bf"}], len = 85
Fri Mar 11 21:48:02 2022 handle_pub: pipe_info size: [0]
Fri Mar 11 21:48:02 2022 server_cb: ERROR it should not happen
Fri Mar 11 21:48:02 2022 nni_aio_finish: Hunky dory(0)
Fri Mar 11 21:48:02 2022 server_cb: END ^^^^ ctx16 ^^^^
Fri Mar 11 21:48:02 2022 nano_pipe_recv_cb: ERROR: pipe is closed abruptly!!
Fri Mar 11 21:48:02 2022 nano_pipe_stop:  ########## nano_pipe_stop ########## 
Fri Mar 11 21:48:02 2022 tcptran_pipe_recv_cb: end of tcptran_pipe_recv_cb: synch! 0x6130000202c0

Fri Mar 11 21:48:02 2022 server_cb: total pipes: 0
Fri Mar 11 21:48:02 2022 nano_pipe_fini:  ########## nano_pipe_fini ########## 
Fri Mar 11 21:48:02 2022 init_pipe_content: pub_handler: init pipe_info
Fri Mar 11 21:48:02 2022 decode_pub_message: cmd: 3, retain: 0, qos: 0, dup: 0, remaining length: 13
Fri Mar 11 21:48:02 2022 decode_pub_message: topic: [will_topic], qos: 0
Fri Mar 11 21:48:02 2022 decode_pub_message: used pos: [12]
Fri Mar 11 21:48:02 2022 decode_pub_message: payload: [], len = 1
Fri Mar 11 21:48:02 2022 handle_pub: pipe_info size: [0]
Fri Mar 11 21:48:02 2022 nni_aio_finish: Hunky dory(0)
Fri Mar 11 21:48:02 2022 conn_param_free: destroy conn param
Fri Mar 11 21:48:02 2022 server_cb: WAIT ^^^^ ctx16 ^^^^
Fri Mar 11 21:48:02 2022 server_cb: total pipes: 0
Fri Mar 11 21:48:02 2022 free_pub_packet: free topic
Fri Mar 11 21:48:02 2022 free_pub_packet: free payload
Fri Mar 11 21:48:02 2022 free_pub_packet: free pub_packet
Fri Mar 11 21:48:02 2022 init_pipe_content: pub_handler: init pipe_info
Fri Mar 11 21:48:02 2022 nng_ctx_recv:  ######## nng_ctx_recv context id 16 ######## 
Fri Mar 11 21:48:02 2022 nano_ctx_recv: nano_ctx_recv start 0x6100000010b0
^C
Broker exit(0).

=================================================================
==2313152==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 256 byte(s) in 4 object(s) allocated from:
    #0 0x7ff88a400dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x5585b80c4e33 in nni_zalloc ../nng/src/platform/posix/posix_alloc.c:26
    #2 0x5585b80950f2 in nng_zalloc ../nng/src/nng.c:64
    #3 0x5585b804af8d in handle_pub ../nanomq/pub_handler.c:159
    #4 0x5585b8078102 in server_cb ../nanomq/apps/broker.c:265
    #5 0x5585b80bfc09 in nni_task_exec ../nng/src/core/taskq.c:144
    #6 0x5585b80a3cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #7 0x5585b80a3fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #8 0x5585b80e0adc in nano_pipe_close ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:776
    #9 0x5585b80b3b69 in nni_pipe_close ../nng/src/core/pipe.c:137
    #10 0x5585b80e2590 in nano_pipe_recv_cb ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:957
    #11 0x5585b80bfc09 in nni_task_exec ../nng/src/core/taskq.c:144
    #12 0x5585b80a3cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #13 0x5585b80a3fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #14 0x5585b81594f8 in tcptran_pipe_recv_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:738
    #15 0x5585b80bf412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #16 0x5585b80c09ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #17 0x5585b80c91e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #18 0x7ff88a2da58f in start_thread nptl/pthread_create.c:463

Direct leak of 128 byte(s) in 2 object(s) allocated from:
    #0 0x7ff88a400dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x5585b80c4e33 in nni_zalloc ../nng/src/platform/posix/posix_alloc.c:26
    #2 0x5585b80950f2 in nng_zalloc ../nng/src/nng.c:64
    #3 0x5585b810aeb8 in property_alloc ../nng/src/supplemental/mqtt/mqtt_codec.c:1939
    #4 0x5585b810b466 in property_set_value_str ../nng/src/supplemental/mqtt/mqtt_codec.c:2056
    #5 0x5585b80d6c11 in conn_handler ../nng/src/sp/protocol/mqtt/mqtt_parser.c:587
    #6 0x5585b8155192 in tcptran_pipe_nego_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:337
    #7 0x5585b80bf412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #8 0x5585b80c09ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #9 0x5585b80c91e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #10 0x7ff88a2da58f in start_thread nptl/pthread_create.c:463

Indirect leak of 344 byte(s) in 4 object(s) allocated from:
    #0 0x7ff88a400dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x5585b80c4e33 in nni_zalloc ../nng/src/platform/posix/posix_alloc.c:26
    #2 0x5585b80950f2 in nng_zalloc ../nng/src/nng.c:64
    #3 0x5585b804f6bb in decode_pub_message ../nanomq/pub_handler.c:578
    #4 0x5585b804afd8 in handle_pub ../nanomq/pub_handler.c:162
    #5 0x5585b8078102 in server_cb ../nanomq/apps/broker.c:265
    #6 0x5585b80bfc09 in nni_task_exec ../nng/src/core/taskq.c:144
    #7 0x5585b80a3cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #8 0x5585b80a3fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #9 0x5585b80e0adc in nano_pipe_close ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:776
    #10 0x5585b80b3b69 in nni_pipe_close ../nng/src/core/pipe.c:137
    #11 0x5585b80e2590 in nano_pipe_recv_cb ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:957
    #12 0x5585b80bfc09 in nni_task_exec ../nng/src/core/taskq.c:144
    #13 0x5585b80a3cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #14 0x5585b80a3fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #15 0x5585b81594f8 in tcptran_pipe_recv_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:738
    #16 0x5585b80bf412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #17 0x5585b80c09ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #18 0x5585b80c91e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #19 0x7ff88a2da58f in start_thread nptl/pthread_create.c:463

Indirect leak of 104 byte(s) in 4 object(s) allocated from:
    #0 0x7ff88a400bc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x5585b80c4e02 in nni_alloc ../nng/src/platform/posix/posix_alloc.c:20
    #2 0x5585b80950d4 in nng_alloc ../nng/src/nng.c:58
    #3 0x5585b80d41c8 in copy_utf8_str ../nng/src/sp/protocol/mqtt/mqtt_parser.c:253
    #4 0x5585b804e692 in decode_pub_message ../nanomq/pub_handler.c:514
    #5 0x5585b804afd8 in handle_pub ../nanomq/pub_handler.c:162
    #6 0x5585b8078102 in server_cb ../nanomq/apps/broker.c:265
    #7 0x5585b80bfc09 in nni_task_exec ../nng/src/core/taskq.c:144
    #8 0x5585b80a3cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #9 0x5585b80a3fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #10 0x5585b80e0adc in nano_pipe_close ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:776
    #11 0x5585b80b3b69 in nni_pipe_close ../nng/src/core/pipe.c:137
    #12 0x5585b80e2590 in nano_pipe_recv_cb ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:957
    #13 0x5585b80bfc09 in nni_task_exec ../nng/src/core/taskq.c:144
    #14 0x5585b80a3cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #15 0x5585b80a3fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #16 0x5585b81594f8 in tcptran_pipe_recv_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:738
    #17 0x5585b80bf412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #18 0x5585b80c09ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #19 0x5585b80c91e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #20 0x7ff88a2da58f in start_thread nptl/pthread_create.c:463

SUMMARY: AddressSanitizer: 832 byte(s) leaked in 14 allocation(s).
```

Test command:
```bash
mqtt sub -t "topic" -d 
```
```bash
mqtt pub -t "topic" -m "hello" -pf "UTF_8" -e 60 -up "user=wh" -ct "BIN" -Cup "alvin=value"  -q 2  -v  --willTopic "will_topic" --willMessageExpiryInterval 180 --willPayloadFormatIndicator "UTF_8" 
```

File: nanomq/sub_handler.c (hunk 2 above)
Comment (alvin1221) [reply nanomq/sub_handler.c:294]: `r` ?

**Comment by alvin1221** [comment]
memory leaks:
```bash
mosquitto_pub -t "topic" -m "hello"  --will-topic "will_topic" --will-payload "will_msg" -d -V 5
```

```bash
Sat Mar 12 16:32:41 2022 nano_pipe_fini:  ########## nano_pipe_fini ########## 
Sat Mar 12 16:32:41 2022 init_pipe_content: pub_handler: init pipe_info
Sat Mar 12 16:32:41 2022 decode_pub_message: cmd: 3, retain: 0, qos: 0, dup: 0, remaining length: 21
Sat Mar 12 16:32:41 2022 decode_pub_message: topic: [will_topic], qos: 0
Sat Mar 12 16:32:41 2022 decode_pub_message: used pos: [12]
Sat Mar 12 16:32:41 2022 decode_pub_message: payload: [], len = 9
Sat Mar 12 16:32:41 2022 handle_pub: pipe_info size: [0]
Sat Mar 12 16:32:41 2022 nni_aio_finish: Hunky dory(0)
Sat Mar 12 16:32:41 2022 conn_param_free: destroy conn param
Sat Mar 12 16:32:41 2022 server_cb: WAIT ^^^^ ctx3 ^^^^
Sat Mar 12 16:32:41 2022 server_cb: total pipes: 0
Sat Mar 12 16:32:41 2022 free_pub_packet: free topic
Sat Mar 12 16:32:41 2022 free_pub_packet: free payload
Sat Mar 12 16:32:41 2022 free_pub_packet: free pub_packet
Sat Mar 12 16:32:41 2022 init_pipe_content: pub_handler: init pipe_info
Sat Mar 12 16:32:41 2022 nng_ctx_recv:  ######## nng_ctx_recv context id 3 ######## 
Sat Mar 12 16:32:41 2022 nano_ctx_recv: nano_ctx_recv start 0x6100000003b0
^C
Broker exit(0).

=================================================================
==2467128==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7f075b234dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x564165068e33 in nni_zalloc ../nng/src/platform/posix/posix_alloc.c:26
    #2 0x5641650390f2 in nng_zalloc ../nng/src/nng.c:64
    #3 0x564164feef8d in handle_pub ../nanomq/pub_handler.c:159
    #4 0x56416501c102 in server_cb ../nanomq/apps/broker.c:265
    #5 0x564165063c09 in nni_task_exec ../nng/src/core/taskq.c:144
    #6 0x564165047cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #7 0x564165047fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #8 0x564165084b57 in nano_pipe_close ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:776
    #9 0x564165057b69 in nni_pipe_close ../nng/src/core/pipe.c:137
    #10 0x56416508660b in nano_pipe_recv_cb ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:957
    #11 0x564165063c09 in nni_task_exec ../nng/src/core/taskq.c:144
    #12 0x564165047cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #13 0x564165047fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #14 0x5641650fd573 in tcptran_pipe_recv_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:738
    #15 0x564165063412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #16 0x5641650649ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #17 0x56416506d1e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #18 0x7f075b10e58f in start_thread nptl/pthread_create.c:463

Indirect leak of 86 byte(s) in 1 object(s) allocated from:
    #0 0x7f075b234dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x564165068e33 in nni_zalloc ../nng/src/platform/posix/posix_alloc.c:26
    #2 0x5641650390f2 in nng_zalloc ../nng/src/nng.c:64
    #3 0x564164ff36bb in decode_pub_message ../nanomq/pub_handler.c:578
    #4 0x564164feefd8 in handle_pub ../nanomq/pub_handler.c:162
    #5 0x56416501c102 in server_cb ../nanomq/apps/broker.c:265
    #6 0x564165063c09 in nni_task_exec ../nng/src/core/taskq.c:144
    #7 0x564165047cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #8 0x564165047fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #9 0x564165084b57 in nano_pipe_close ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:776
    #10 0x564165057b69 in nni_pipe_close ../nng/src/core/pipe.c:137
    #11 0x56416508660b in nano_pipe_recv_cb ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:957
    #12 0x564165063c09 in nni_task_exec ../nng/src/core/taskq.c:144
    #13 0x564165047cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #14 0x564165047fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #15 0x5641650fd573 in tcptran_pipe_recv_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:738
    #16 0x564165063412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #17 0x5641650649ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #18 0x56416506d1e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #19 0x7f075b10e58f in start_thread nptl/pthread_create.c:463

Indirect leak of 26 byte(s) in 1 object(s) allocated from:
    #0 0x7f075b234bc8 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x564165068e02 in nni_alloc ../nng/src/platform/posix/posix_alloc.c:20
    #2 0x5641650390d4 in nng_alloc ../nng/src/nng.c:58
    #3 0x5641650781c8 in copy_utf8_str ../nng/src/sp/protocol/mqtt/mqtt_parser.c:253
    #4 0x564164ff2692 in decode_pub_message ../nanomq/pub_handler.c:514
    #5 0x564164feefd8 in handle_pub ../nanomq/pub_handler.c:162
    #6 0x56416501c102 in server_cb ../nanomq/apps/broker.c:265
    #7 0x564165063c09 in nni_task_exec ../nng/src/core/taskq.c:144
    #8 0x564165047cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #9 0x564165047fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #10 0x564165084b57 in nano_pipe_close ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:776
    #11 0x564165057b69 in nni_pipe_close ../nng/src/core/pipe.c:137
    #12 0x56416508660b in nano_pipe_recv_cb ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:957
    #13 0x564165063c09 in nni_task_exec ../nng/src/core/taskq.c:144
    #14 0x564165047cba in nni_aio_finish_impl ../nng/src/core/aio.c:450
    #15 0x564165047fa6 in nni_aio_finish_sync ../nng/src/core/aio.c:467
    #16 0x5641650fd573 in tcptran_pipe_recv_cb ../nng/src/sp/transport/mqtt/broker_tcp.c:738
    #17 0x564165063412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #18 0x5641650649ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #19 0x56416506d1e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #20 0x7f075b10e58f in start_thread nptl/pthread_create.c:463

SUMMARY: AddressSanitizer: 176 byte(s) leaked in 3 allocation(s).
```

**Comment by alvin1221** [comment]
crash
```bash
mosquitto_pub -t "topic" -m "hello"  --will-topic "will_topic" --will-payload "will_msg" 
```

```bash
Sat Mar 12 16:32:14 2022 tcptran_pipe_recv_cb: end of tcptran_pipe_recv_cb: synch! 0x61300000ffc0

Sat Mar 12 16:32:14 2022 server_cb: total pipes: 0
Sat Mar 12 16:32:14 2022 nano_pipe_fini:  ########## nano_pipe_fini ########## 
Sat Mar 12 16:32:14 2022 init_pipe_content: pub_handler: init pipe_info
Sat Mar 12 16:32:14 2022 decode_pub_message: cmd: 3, retain: 0, qos: 0, dup: 0, remaining length: 2
Sat Mar 12 16:32:14 2022 decode_pub_message: topic: [], qos: 0
Sat Mar 12 16:32:14 2022 decode_pub_message: used pos: [2]
[ERROR] (../nanolib/mqtt_db.c:993: errno: None) db or topic is NULL
[ERROR] (../nanolib/mqtt_db.c:103: errno: None) topic is NULL
AddressSanitizer:DEADLYSIGNAL
=================================================================
==2466858==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x56541b68fb57 bp 0x7fa325bf4780 sp 0x7fa325bf4620 T11)
==2466858==The signal is caused by a READ memory access.
==2466858==Hint: address points to the zero page.
    #0 0x56541b68fb56 in search_client ../nanolib/mqtt_db.c:1016
    #1 0x56541b68ff3f in dbtree_find_clients_and_cache_msg ../nanolib/mqtt_db.c:1055
    #2 0x56541b6b4537 in handle_pub ../nanomq/pub_handler.c:200
    #3 0x56541b6e5c31 in server_cb ../nanomq/apps/broker.c:575
    #4 0x56541b728412 in nni_taskq_thread ../nng/src/core/taskq.c:50
    #5 0x56541b7299ea in nni_thr_wrap ../nng/src/core/thread.c:94
    #6 0x56541b7321e4 in nni_plat_thr_main ../nng/src/platform/posix/posix_thread.c:266
    #7 0x7fa32e58e58f in start_thread nptl/pthread_create.c:463
    #8 0x7fa32e4a9222 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x119222)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV ../nanolib/mqtt_db.c:1016 in search_client
Thread T11 (nng:task) created by T0 here:
    #0 0x7fa32e5e1805 in pthread_create (/lib/x86_64-linux-gnu/libasan.so.5+0x3a805)
    #1 0x56541b732314 in nni_plat_thr_init ../nng/src/platform/posix/posix_thread.c:279
    #2 0x56541b729c93 in nni_thr_init ../nng/src/core/thread.c:121
    #3 0x56541b728733 in nni_taskq_init ../nng/src/core/taskq.c:95
    #4 0x56541b7296ba in nni_taskq_sys_init ../nng/src/core/taskq.c:284
    #5 0x56541b714773 in nni_init_helper ../nng/src/core/init.c:35
    #6 0x56541b732668 in nni_plat_init ../nng/src/platform/posix/posix_thread.c:422
    #7 0x56541b7147f1 in nni_init ../nng/src/core/init.c:58
    #8 0x56541b74caac in nni_proto_mqtt_open ../nng/src/sp/protocol.c:37
    #9 0x56541b74c80f in nng_nmq_tcp0_open ../nng/src/sp/protocol/mqtt/nmq_mqtt.c:1173
    #10 0x56541b6e6ae3 in broker ../nanomq/apps/broker.c:732
    #11 0x56541b6e9cff in broker_start ../nanomq/apps/broker.c:1165
    #12 0x56541b6b0919 in main ../nanomq/nanomq.c:187
    #13 0x7fa32e3b8cb1 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28cb1)

==2466858==ABORTING
```

**Pull request merged**